rubix-hadoop2/src/main/java/com/qubole/rubix/hadoop2/Hadoop2ClusterManager.java

```diff
@@ -51,89 +39,41 @@ public void initialize(Configuration conf)
-        Thread t = Executors.defaultThreadFactory().newThread(r);
-        t.setName("rubix-get-nodes-thread");
-        t.setDaemon(true);
-        return t;
-      }
-    });
-    nodesCache = CacheBuilder.newBuilder()
-        .refreshAfterWrite(getNodeRefreshTime(), TimeUnit.SECONDS)
-        .build(CacheLoader.asyncReloading(new CacheLoader<String, List<String>>()
-        {
-          @Override
-          public List<String> load(String s)
-              throws Exception
-          {
-            if (!isMaster) {
-              // First time all nodes start assuming themselves as master and down the line figure out their role
-              // Next time onwards, only master will be fetching the list of nodes
-              return ImmutableList.of();
-            }
-            try {
-              List<Hadoop2ClusterManagerUtil.Node> allNodes = Hadoop2ClusterManagerUtil.getAllNodes(yconf);
-              if (allNodes == null) {
-                isMaster = false;
-                return ImmutableList.of();
-              }
-
-              Set<String> hosts = new HashSet<>();
-
-              if (allNodes.isEmpty()) {
-                // Empty result set => server up and only master node running, return localhost has the only node
-                // Do not need to consider failed nodes list as 1node cluster and server is up since it replied to allNodesRequest
-                return ImmutableList.of(InetAddress.getLocalHost().getHostAddress());
-              }
-
-              for (Hadoop2ClusterManagerUtil.Node node : allNodes) {
-                String state = node.getState();
-                log.debug("Hostname: " + node.getNodeHostName() + "State: " + state);
-                //keep only healthy data nodes
-                if (state.equalsIgnoreCase("Running") || state.equalsIgnoreCase("New") || state.equalsIgnoreCase("Rebooted")) {
-                  hosts.add(node.getNodeHostName());
-                }
-              }
-
-              if (hosts.isEmpty()) {
-                throw new Exception("No healthy data nodes found.");
-              }
-
-              List<String> hostList = Lists.newArrayList(hosts.toArray(new String[0]));
-              Collections.sort(hostList);
-              log.debug("Hostlist: " + hostList.toString());
-              return hostList;
-            }
-            catch (Exception e) {
-              throw Throwables.propagate(e);
-            }
-          }
-        }, executor));
-  }
-
-  @Override
-  public boolean isMaster()
-      throws ExecutionException
-  {
-    // issue get on nodesSupplier to ensure that isMaster is set correctly
-    nodesCache.get("nodeList");
-    return isMaster;
   }
 
   @Override
-  public List<String> getNodes()
+  public List<String> getNodesInternal()
   {
     try {
-      return nodesCache.get("nodeList");
+      List<Hadoop2ClusterManagerUtil.Node> allNodes = Hadoop2ClusterManagerUtil.getAllNodes(yconf);
+      if (allNodes == null) {
+        return null;
```

Comment:
we can return an empty list so that the caller doesn't need to do the null check.

Reply:
nevermind just saw the comment related to the distinction of the empty and null cases.